compiler/rustc_lint/src/types.rs

```diff
@@ -169,6 +170,32 @@ declare_lint! {
     "detects ambiguous wide pointer comparisons"
 }
 
+declare_lint! {
+    /// The `unpredictable_function_pointer_comparisons` lint checks comparison
+    /// of function pointer as the operands.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// fn foo() {}
+    /// # let a = foo as fn();
+    ///
+    /// let _ = a == foo;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Function pointers comparisons do not produce meaningful result since
+    /// they are never guaranteed to be unique and could vary between different
+    /// code generation units. Furthermore different function could have the
+    /// same address after being merged together.
```

Comment:
The detailed explanation (and maybe even the diagnostic message) should say that `==` on fn ptrs goes wrong in both ways: 
- different functions can compare equal (because LLVM deduplicates functions that have the same assembly)
- the same function can have multiple addresses

Reply:
Sure, but isn't this already the case? The first sentence seems to be your second bullet point and second sentence seems to be your first bullet point.

Or is your point that it needs to detailed explanation?

Reply:
I've improved the diagnostic output to include a note about function de-duplication.

<details>

```diff
 warning: function pointer comparisons do not produce meaningful results since their addresses are not guaranteed to be unique
   --> $DIR/fn-ptr-comparisons.rs:16:13
    |
 LL |     let _ = f == a;
    |             ^^^^^^
    |
    = note: even foreign function pointers are not guaranteed to be unique and could vary between different code generation units
+   = note: furthermore different functions could have the same address after being merged together
```

</details>

Reply:
Ah fair, you are mentioning them both, sorry.

Maybe it could say more explicitly: the address of the same function can very between different codegen units.

Also, there should be a comma after "furthermore".

I don't understand why "even foreign function pointers" are mentioned?

Reply:
> I don't understand why "even foreign function pointers" are mentioned?

I wanted to emphasise that even functions from other crates or FFIs are also subject to unpredictable comparisons.

But I can now see that it's not clear enough. I will improve the wording.

Reply:
We have very limited space in the diagnostic message, I wouldn't waste that on rare corner cases like FFI. Just focus on Rust functions.

In the lint explanation you can expand on this if you want.

But I think for FFI-imported functions we do actually have reliable comparisons if the target language we link against has reliable comparisons?

Reply:
By "unpredictable comparisons" for FFI, I mean that like Rust functions, they can also be subject to deduplication and different addresses, with LTO and cross-language inlining for example.

Reply:
That's true only of the other language enables those attributes. If you import C functions, as far as I know, you do get reliable comparison.

I don't see why the lint would have to discuss this though.

Reply:
I don't feel strongly about the foreign function part so I removed it. I also applied your suggestion.

<details>

<summary>The new diagnostic</summary>

```diff
-   = note: even foreign function pointers are not guaranteed to be unique and could vary between different code generation units
-   = note: furthermore different functions could have the same address after being merged together
+   = note: the address of the same function can very between different codegen units
+   = note: furthermore, different functions could have the same address after being merged together
```

</details>

Reply:
That sounds good, thanks. :)